### PR TITLE
[EOC-575] Same flow for removing account google and others

### PR DESCRIPTION
**Pull request by adamklepacz**
Zrzut ekranu 2019-10-25 o 08 43 46
Zrzut ekranu 2019-10-25 o 08 43 42
Zrzut ekranu 2019-10-25 o 08 43 38
Zrzut ekranu 2019-10-25 o 08 41 49
Zrzut ekranu 2019-10-25 o 08 41 39
Zrzut ekranu 2019-10-25 o 08 41 21
Zrzut ekranu 2019-10-25 o 08 49 00

### CODE REVIEW

File: client/src/locales/en.json [hunk 1]
```diff
@@ -216,6 +216,11 @@
   "dashboard.index.no-arch-sacks": "There are no archived sacks yet.",
   "dashboard.index.no-sacks": "There are no sacks yet.",
   "dashboard.index.show-arch-sacks": "Show archived sacks",
+  "delete-link.has-expired-part1": "Delete link has expired. Please go to",
+  "delete-link.has-expired-link": "user profile page ",
+  "delete-link.has-expired-part2": "and find delete account option to get new delete link.",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: Do not cut sentences in the middle - use parameter if needed. It is hard or impossible to translate such divided sentences.

File: client/src/locales/en.json [hunk 2]
```diff
@@ -216,6 +216,11 @@
...
+  "email.delete-account.failure": "There was an error while sending email.",
+  "email.delete-account.success": "Email with delete button was sent to your email inbox. To delete your account follow email instructions.",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: Check other services what text is used. Most probably sth like "Delete account confirmation email was sent...". Please check and use more accurate text without some 'button' text.

File: client/src/modules/user/AuthBox/components/DeleteLinkExpired.jsx [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+      <HomeLink />
+    </h1>
+    <p>
+      <FormattedMessage id="delete-link.has-expired-part1" />
+      <span> </span>
+      <Link className="delete-link-expired__link" to="/user-profile">
+        <FormattedMessage id="delete-link.has-expired-link" />
+      </Link>
+      <span> </span>
+      <FormattedMessage id="delete-link.has-expired-part2" />
```
Comment (marekrozmus) [inline client/src/modules/user/AuthBox/components/DeleteLinkExpired.jsx]: use one message with parameters

File: client/src/modules/user/model/actions.js [hunk 4]
```diff
@@ -163,8 +162,26 @@ export const getAccountDetails = token =>
 
 export const checkIfDataLeft = () => getData('/auth/check-if-data-left');
 
-export const deleteAccount = (email, password) =>
-  deleteData('/auth', { email, password });
+export const sendDeleteAccountMail = () => async dispatch => {
+  try {
+    const result = await getData('/auth/send-delete-account-mail');
```
Comment (marekrozmus) [inline client/src/modules/user/model/actions.js]: just `/auth/delete-account`

File: server/common/variables.js [hunk 5]
```diff
@@ -69,6 +69,7 @@ const ActivityType = Object.freeze({
 const LOCK_TIMEOUT = 300000;
 const SOCKET_TIMEOUT = 60000;
 const EXPIRATION_TIME = 3600000;
+const DELETE_MAIL_EXPIRATION_TIME = 3600000;
```
Comment (marekrozmus) [inline server/common/variables.js]: one expiration time for all email links

Comment (alemikolo) [inline server/common/variables.js]: use `EXPIRATION_TIME`

File: server/controllers/userAuth.js [hunk 6]
```diff
@@ -509,8 +509,41 @@ const deleteAccount = async (req, resp) => {
       req.logout();
       resp.clearCookie('connect.sid');
 
-      return resp.send();
+      return resp.redirect('/account-deleted');
     }
+
+    resp.redirect(`/delete-link-expired/${deleteToken}`);
```
Comment (alemikolo) [inline server/controllers/userAuth.js]: why we send the token when it is expired?

File: server/controllers/userAuth.js [hunk 7]
```diff
@@ -509,8 +509,41 @@ const deleteAccount = async (req, resp) => {
...
+
+    if (user) {
+      // eslint-disable-next-line no-param-reassign
+      const data = {
+        deleteToken,
+        email,
+        req
+      };
+
+      mailer.sendDeleteAccountMail(data);
```
Comment (alemikolo) [inline server/controllers/userAuth.js]: is this sync or async?

File: server/migrations/20191024120045-0.1.0.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+/* eslint-disable no-param-reassign */
+const { runAsyncTasks } = require('../common/utils');
+
+const up = async db => {
+  const addDeleteTokenData = async () =>
+    db.collection('users').updateMany(
+      {
+        deleteToken: { $exists: false },
```
Comment (alemikolo) [inline server/migrations/20191024120045-0.1.0.js]: it's not necessary to add these fields to all users. Refresh and confirmation tokens are added when they are needed. We add these fields when user decides to remove account. Why litter the document with redundant data?

File: client/src/modules/user/model/actions.js (hunk 4 above)
Comment (adamklepacz) [reply client/src/modules/user/model/actions.js]: I can't.
 
1. This action is not deleting account but sending email with delete link
2. /delete-account API endpoint is allready in use for a controller thath is deleting account
Zrzut ekranu 2019-10-25 o 10 35 58

File: server/migrations/20191024120045-0.1.0.js (hunk 8 above)
Comment (adamklepacz) [reply server/migrations/20191024120045-0.1.0.js]: so is the migration needed in this case or not?

File: client/src/locales/en.json [hunk 9]
```diff
@@ -216,6 +216,10 @@
   "dashboard.index.no-arch-sacks": "There are no archived sacks yet.",
   "dashboard.index.no-sacks": "There are no sacks yet.",
   "dashboard.index.show-arch-sacks": "Show archived sacks",
+  "delete-link.has-expired": "Delete link has expired. Please go to {link} and find delete account option to get new delete link.",
+  "delete-link.has-expired-link": "user profile page ",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: remove space at the end

File: client/src/locales/en.json [hunk 10]
```diff
@@ -216,6 +216,10 @@
   "dashboard.index.no-arch-sacks": "There are no archived sacks yet.",
   "dashboard.index.no-sacks": "There are no sacks yet.",
   "dashboard.index.show-arch-sacks": "Show archived sacks",
+  "delete-link.has-expired": "Delete link has expired. Please go to {link} and find delete account option to get new delete link.",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: "Delete account link has expired. If you still want to delete account go to user profile page."

File: server/controllers/userAuth.js [hunk 11]
```diff
@@ -509,8 +508,40 @@ const deleteAccount = async (req, resp) => {
...
+    const deleteToken = crypto.randomBytes(32).toString('hex');
+    const deleteTokenExpirationDate = new Date().getTime() + EXPIRATION_TIME;
+
+    const user = await User.findOneAndUpdate(
+      { _id },
+      { deleteToken, deleteTokenExpirationDate }
+    );
+
+    if (user) {
+      // eslint-disable-next-line no-param-reassign
```
Comment (marekrozmus) [inline server/controllers/userAuth.js]: why this rule is disabled here?

Comment (adamklepacz) [reply server/controllers/userAuth.js]: legacy code

File: client/src/locales/en.json [hunk 12]
```diff
@@ -216,6 +216,10 @@
   "dashboard.index.no-arch-sacks": "There are no archived sacks yet.",
   "dashboard.index.no-sacks": "There are no sacks yet.",
   "dashboard.index.show-arch-sacks": "Show archived sacks",
+  "delete-link.has-expired": "Delete account link has expired. If you still want to delete account go to {link} page.",
+  "delete-link.has-expired-link": "user profile page",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: If this one is used in previous text then you'll get double `page` word at the end of the sentence.

File: client/src/locales/en.json [hunk 13]
```diff
@@ -216,6 +216,10 @@
...
+  "email.delete-account.failure": "There was an error while sending email.",
+  "email.delete-account.success": "An email containing a link that will allow you to delete your account has been sent.",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: I think that 'allow' word might not be appropriate. Maybe just change to: 'We sent you an email with instructions to delete your account.' or 'Check your email for a link to delete your account. If it doesn’t appear within a few minutes, check your spam folder.'

File: server/controllers/userAuth.js [hunk 14]
```diff
@@ -509,8 +508,39 @@ const deleteAccount = async (req, resp) => {
...
+        email,
+        req
+      };
+
+      await mailer.sendDeleteAccountMail(data);
+
+      return resp.sendStatus(200);
     }
+
     throw new Error();
```
Comment (marekrozmus) [inline server/controllers/userAuth.js]: Please change that code - do not throw error just try sth like this:
```
if (!user) {
  return resp.sendStatus(400);
}

const data = {
  deleteToken,
  email,
  req
};

await mailer.sendDeleteAccountMail(data);

return resp.sendStatus(200);
```

File: client/src/locales/en.json [hunk 15]
```diff
@@ -216,6 +216,10 @@
   "dashboard.index.no-arch-sacks": "There are no archived sacks yet.",
   "dashboard.index.no-sacks": "There are no sacks yet.",
   "dashboard.index.show-arch-sacks": "Show archived sacks",
+  "delete-link.has-expired": "Delete account link has expired. If you still want to delete account go to {link}",
+  "delete-link.has-expired-link": "user profile page.",
```
Comment (marekrozmus) [inline client/src/locales/en.json]: move ` page.` to `has-expired` text

**Pull request merged**